### PR TITLE
added a way to executing analyzes from plugin and to customize reports

### PR DESCRIPTION
**Pull request by Halleck45**
Please see an exampe of plugin here: [symfony2extension](https://github.com/phpmetrics/SymfonyExtension)

### CODE REVIEW

**Comment by Halleck45** [comment]
Another exemple: [composer extension](https://github.com/phpmetrics/ComposerExtension)

File: src/Hal/Application/Command/RunMetricsCommand.php [hunk 1]
```diff
@@ -116,19 +123,45 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+            }
+            $plugin = require_once($filename);
+            $repository->attach($plugin);
+        }
+        $extensionService = new ExtensionService($repository);
+
         // prepare structures
         $bounds = new Bounds();
         $collection = new \Hal\Component\Result\ResultCollection();
         $aggregatedResults = new \Hal\Component\Result\ResultCollection();
```
Comment (UFOMelkor) [inline src/Hal/Application/Command/RunMetricsCommand.php]: Can be imported both

File: src/Hal/Application/Config/Configuration.php [hunk 2]
```diff
@@ -55,6 +55,11 @@ class Configuration implements ConfigurationInterface
     private $ignoreErrors = false;
 
     /**
+     * @var array
+     */
+    private $extensions = array();
```
Comment (UFOMelkor) [inline src/Hal/Application/Config/Configuration.php]: Within the constructor you set `$this->extensions = new ExtensionsConfiguration();`

File: src/Hal/Application/Extension/ExtensionService.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    public function __construct(Repository $repository)
+    {
+        $this->repository = $repository;
+    }
+
+    /**
+     * @return Repository
+     */
+    public function getRepository()
```
Comment (UFOMelkor) [inline src/Hal/Application/Extension/ExtensionService.php]: Every time this method is called the call looks like `$extensionService->getRepository()->all()`. We might consider replacing `getRepository()` with a `getAllExtensions()` method.

File: src/Hal/Application/Formater/Details/Json.php [hunk 4]
```diff
@@ -26,13 +27,20 @@ class Json implements FormaterInterface {
...
+    private $extensionsService;
+
+    /**
      * Constructor
      *
      * @param boolean $prettyPrint optional pretty printing for result json
+     * @param ExtensionService $extensionService
      */
-    public function __construct($prettyPrint = false)
+    public function __construct($prettyPrint = false, ExtensionService $extensionService)
```
Comment (UFOMelkor) [inline src/Hal/Application/Formater/Details/Json.php]: Optional parameter before required one

File: src/Hal/Application/Command/Job/QueueAnalyzeFactory.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Factory queue
+     *
+     * @param Finder $finder
+     * @param BoundsInterface $bounds
+     * @return Queue
+     */
+    public function factory(Finder $finder, BoundsInterface $bounds) {
+        $rules = $this->config->getRuleSet();
+        $validator = new \Hal\Application\Rule\Validator($rules);
```
Comment (UFOMelkor) [inline src/Hal/Application/Command/Job/QueueAnalyzeFactory.php]: Can be imported

File: src/Hal/Application/Config/Configuration.php [hunk 6]
```diff
@@ -173,5 +179,22 @@ public function setIgnoreErrors($ignoreErrors)
...
+    /**
+     * @return array
+     */
+    public function getExtensions()
+    {
+        return $this->extensions;
+    }
+
+    /**
+     * @param array $extensions
```
Comment (UFOMelkor) [inline src/Hal/Application/Config/Configuration.php]: `@param ExtensionsConfiguration $extensions`

File: src/Hal/Application/Extension/ExtensionService.php (hunk 3 above)
Comment (Halleck45) [inline src/Hal/Application/Extension/ExtensionService.php]: I've changed to code according your other comments.

I prefer avoid methods like `getAllExtensions()`, `getFavoriteXXX()`, `getFirtXXX()`... That's useful, but, according to me, using these "shortcuts" methods increase complexity of code and transform "logical code" into "tool code".

**Pull request merged**